### ISSUE TITLE
InsertTokensBefore has non-obvious restriction

### ISSUE THREAD

**Issue by ghost** [dotnet/roslyn#3082]
I wrote a fixer like this, to mark a class as static:

```
    private async Task<Document> MarkClassAsStatic(Document document, ClassDeclarationSyntax classDeclaration, CancellationToken c)
    {
        SyntaxToken classDeclarationToken = classDeclaration.Keyword;
        SyntaxToken staticKeywordToken = SyntaxFactory.Token(SyntaxKind.StaticKeyword);

        SyntaxNode root = await document.GetSyntaxRootAsync();
        SyntaxNode newRoot = root.InsertTokensBefore(classDeclarationToken, new[] { staticKeywordToken });

        Document newDocument = document.WithSyntaxRoot(newRoot);
        return newDocument;
    }
```

The call to InsertTokensBefore fails with the error "System.InvalidOperationException: The item specified is not the element of a list."

It turns out that the first argument to InsertTokensBefore must be an element of a list, and the "class" keyword is not an element of a list. In fact that's why the parameter name is tokenInList. Intellisense shows that, but when I saw the Intellisense for this method I had no idea what the name meant.

And it turns out there's a direct way to do what I want, using the AddModifiers method:

```
        ClassDeclarationSyntax newClassDeclaration = classDeclaration.AddModifiers(new SyntaxToken[] { staticKeywordToken });

        SyntaxNode root = await document.GetSyntaxRootAsync();
        SyntaxNode newRoot = root.ReplaceNode(classDeclaration, newClassDeclaration);
```

Still, InsertTokensBefore was so seductive, it was really not obvious to me that it wouldn't work, and the error message didn't lead me towards the right answer.

**Comment by gafter** [comment]
I'm not sure what would be an appropriate resolution. Do you think the API documentation should be improved?

**Comment by lgolding** [comment]
I think that a more verbose error message on the exception would help. Something to the effect of "you can't pass an arbitrary token as the first argument. The token has to be a member of a SyntaxTokenList."

**Comment by Logerfo** [comment]
Stumbled into this today, that's really confusing. Thank you @lgolding for providing a solution, helped a lot!

**Comment by cezarypiatek** [comment]
So how to deal with scenario when complete replacement is required? I've described my problem in details here https://stackoverflow.com/questions/48953387/replace-single-assignment-statement-with-multiple-statements-with-roslyn

**Comment by rtnlmeme-DestroyerOfDeath** [comment]
Ran into the same thing with InsertNodesBefore. The docstring says "an element of a list" but if I just start out with the api I'm just reading over this and not realizing that the meaning is "You can only use this method on a node that is part of a syntax list (e.g. ArgumentSyntaxList)". I just assume that this method inserts nodes in the tree in before the given node. You should throw a better error and put a better docstring.

**Comment by CyrusNajmabadi** [comment]
@rtnlmeme-DestroyerOfDeath 

> You should throw a better error and put a better docstring.

PRs welcome :-)

**Comment by magiak** [comment]
I still don't get it :( I have UsingDirectiveSyntax and I would like to add global keyword token to it. Can I use InsertTokensBefore?

**Comment by CyrusNajmabadi** [comment]
> I still don't get it :( I have UsingDirectiveSyntax and I would like to add global keyword token to it. Can I use InsertTokensBefore?

No, you cannot.  InsertTokensBefore is only for updating a SyntaxTokenList (like the modifiers-list on a MethodDeclarationSyntax).

That's not how UsingDirectiveSyntax works.  It doesn't have a SYntaxTokenList in it.  It either has a direct GlobalKeyword token or not.  If you want to add a global keyword, just take the UisngDirective and call .Update on it, passing in an appropraite new token for that space.
